src/main/java/org/elasticsearch/action/support/replication/TransportShardReplicationOperationAction.java

```diff
@@ -302,149 +301,159 @@ public void writeTo(StreamOutput out) throws IOException {
         }
     }
 
-    protected class AsyncShardOperationAction {
+    /**
+     * Responsible for performing all operations up to the point we start starting sending requests to replica shards.
+     * Including forwarding the request to another node if the primary is not assigned locally.
+     * <p/>
+     * Note that as soon as we start sending request to replicas, state responsibility is transferred to {@link ReplicationPhase}
+     */
+    final class PrimaryPhase {
         private final ActionListener<Response> listener;
         private final InternalRequest internalRequest;
-        private volatile ShardIterator shardIt;
-        private final AtomicBoolean primaryOperationStarted = new AtomicBoolean();
-        private volatile ClusterStateObserver observer;
+        private final ClusterStateObserver observer;
+        private final AtomicBoolean finished = new AtomicBoolean(false);
 
-        AsyncShardOperationAction(Request request, ActionListener<Response> listener) {
+
+        PrimaryPhase(Request request, ActionListener<Response> listener) {
             this.internalRequest = new InternalRequest(request);
             this.listener = listener;
+            this.observer = new ClusterStateObserver(clusterService, internalRequest.request().timeout(), logger);
         }
 
         public void start() {
-            this.observer = new ClusterStateObserver(clusterService, internalRequest.request().timeout(), logger);
-            doStart();
+            try {
```

Comment:
I changed it to be AbstractRunnable for other reasons.. the class is package private now